common/functions/src/aggregates/aggregate_retention.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::alloc::Layout;
+use std::fmt;
+use std::sync::Arc;
+
+use bytes::BytesMut;
+use common_datavalues::prelude::*;
+use common_exception::ErrorCode;
+use common_exception::Result;
+use common_io::prelude::*;
+use serde_json::json;
+use serde_json::Value as JsonValue;
+
+use super::aggregate_function::AggregateFunction;
+use super::aggregate_function::AggregateFunctionRef;
+use super::aggregate_function_factory::AggregateFunctionDescription;
+use super::StateAddr;
+use crate::aggregates::aggregator_common::assert_variadic_arguments;
+
+struct AggregateRetentionState {
+    pub events: u32,
+}
+
+impl AggregateRetentionState {
+    #[inline(always)]
+    fn add(&mut self, event: u8) {
+        self.events |= 1 << event;
+    }
+
+    fn merge(&mut self, other: &AggregateRetentionState) {
+        self.events |= other.events;
+    }
+
+    fn serialize(&self, writer: &mut BytesMut) -> Result<()> {
+        serialize_into_buf(writer, &self.events)
+    }
+
+    fn deserialize(&mut self, reader: &mut &[u8]) -> Result<()> {
+        self.events = deserialize_from_slice(reader)?;
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+pub struct AggregateRetentionFunction {
+    display_name: String,
+    events_size: u8,
```

Comment:
`events` is `u32`, what will happen if `event_size` is large than `32`?

Reply:
There are validations in try_create

Reply:
Yes, in function `try_create_aggregate_retention_function`, validation code:

```
assert_variadic_arguments(display_name, arguments.len(), (1, 32))?;
```

This assert litmits the events_size from 1 to 32